lib/sgtn-client/common.rb

```diff
@@ -0,0 +1,27 @@
+# Copyright 2022 VMware, Inc.
+# SPDX-License-Identifier: EPL-2.0
+
+module SgtnClient
+  class SingleOperation
+    def initialize(*conditions, &block)
+      raise 'no way to create a new obj' unless block
+
+      @lock = Mutex.new
+      @hash = {}
+
+      @conditions = conditions
+      @creator = block
+    end
+
+    # return new created object
+    def single_operation(id, *args)
+      @lock.synchronize do
+        obj = @hash[id]
+        @conditions.each do |con|
+          return obj unless con.call(id, obj, *args)
+        end
+        @hash[@id] = @creator.call(id, obj, *args)
```

Comment:
```suggestion
        @hash[id] = @creator.call(id, obj, *args)
```

Reply:
This error was not caught in unit test. It basically adds new thread as @hash[nil], which is also used for next calling thread. Could you update unit test to catch this error?

Reply:
I think this will also solve the error you saw with get_cs on source bundle.

> https://github.com/vmware/singleton/runs/5508656880?check_suite_focus=true
> ThreadError:
>       ` Target thread must not be current thread`

Reply:
Same error after change.